src/bucket/BucketSnapshotManager.cpp

```diff
@@ -61,9 +64,13 @@ BucketSnapshotManager::recordBulkLoadMetrics(std::string const& label,
 
 void
 BucketSnapshotManager::maybeUpdateSnapshot(
-    std::unique_ptr<BucketListSnapshot const>& snapshot) const
+    std::unique_ptr<BucketListSnapshot const>& snapshot,
+    std::map<uint32_t, std::unique_ptr<BucketListSnapshot const>>&
+        historicalSnapshots) const
 {
     std::lock_guard<std::recursive_mutex> lock(mSnapshotMutex);
```

Comment:
Why do we need an exclusive lock here? I would expect this and all the other read-only methods to be using `std::shared_lock`. Otherwise you're blocking all the threads on every read unnecessarily.

I haven't looked at this code closely before, and I was a bit confused by it. It's a bit weird that we copy snapshots on every access instead of accessing them directly. My intuition would be that for something like this we would either just use the shared locks for readers directly (e.g. lock in snapshot manager, pass a callback from the reader to read from the snapshot), or return a shared ptr (though I suspect this is slower for our use case since snapshots are not updated frequently). In any case, it doesn't seem like there is a need to use exclusive locks for most of the operations. Am I missing something here?

Reply:
Sorry, the snapshot interface is a little convoluted, but we do directly access snapshots.

Essentially, the `BucketSnapshotManager` holds the canonical, up-to-date BucketList snapshot (`BucketSnapshotManager::mCurrentSnapshot`). Whenever `addBatch` is called on the BucketList, this canonical snapshot is updated via `BucketSnapshotManager::updateCurrentSnapshot`.

The expected interface is that a thread will request a snapshot from the manager via `getSearchableBucketListSnapshot`. This will copy the snapshot over to the local thread. The reason we copy is that the snapshot contains a pointer to a const `Bucket` and a file stream per-bucket. We always copy snapshots such that each thread maintains it's own file streams.

On each query to a snapshot, the snapshot first checks with the `BucketSnapshotManager` to see if it still up to date via `BucketSnapshotManager::maybeUpdateSnapshot`. If the snapshot is up to date, nothing happens. If the snapshot is out of date, the snapshot object updates itself via copy from the canonical snapshot in the `BucketSnapshotManager.`

We acquire this lock in the snapshot manager when either:
1. `addBatch` is called and the canonical snapshot needs to be updated
2. a thread is checking their local snapshot copy against the canonical snapshot and potentially updates via copy.

We could change this to a shared lock where case 1 requires a unique_lock and case 2 requires a shared_lock. I don't think this is a huge issue though. The reason we have threads maintain a local copy with local filestreams is so we don't have to lock on actual reads, we only lock to do a quick comparison of integers then release the lock. Even when we do need to get a new snapshot, the snapshot class is only an integer and a handful of pointers, so the copy is lightweight. TLDR we lock to check if the snapshot is out of date, but release before doing the actual reads.

Reply:
Yeah, I think I've got the idea from the code mostly, not sure if this could be simplified, but ultimately my main point here was regarding shared/exclusive locks - we have orders of magnitude more reads than writes and it's weird to block reads on each other, even for a short period of time. So unless this is proven to be slower by the benchmarks, I'd go for an option that is more correct semantically. Thanks for the update.